File: modules/api/pkg/provider/cloud/packet/provider.go
/*
Copyright 2020 The Kubermatic Kubernetes Platform contributors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package packet

import (
	"errors"
	"fmt"

	"github.com/packethost/packngo"

	"k8c.io/dashboard/v2/pkg/provider"
	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
	"k8c.io/kubermatic/v2/pkg/resources"
)

type packet struct {
	secretKeySelector provider.SecretKeySelectorValueFunc
}

// NewCloudProvider creates a new packet provider.
func NewCloudProvider(secretKeyGetter provider.SecretKeySelectorValueFunc) provider.CloudProvider {
	return &packet{
		secretKeySelector: secretKeyGetter,
	}
}

var _ provider.CloudProvider = &packet{}

func GetCredentialsForCluster(cloudSpec kubermaticv1.CloudSpec, secretKeySelector provider.SecretKeySelectorValueFunc) (apiKey, projectID string, err error) {
	apiKey = cloudSpec.Packet.APIKey
	projectID = cloudSpec.Packet.ProjectID

	if apiKey == "" {
		if cloudSpec.Packet.CredentialsReference == nil {
			return "", "", errors.New("no credentials provided")
		}
		apiKey, err = secretKeySelector(cloudSpec.Packet.CredentialsReference, resources.PacketAPIKey)
		if err != nil {
			return "", "", err
		}
	}

	if projectID == "" {
		if cloudSpec.Packet.CredentialsReference == nil {
			return "", "", errors.New("no credentials provided")
		}
		projectID, err = secretKeySelector(cloudSpec.Packet.CredentialsReference, resources.PacketProjectID)
		if err != nil {
			return "", "", err
		}
	}

	return apiKey, projectID, nil
}

func ValidateCredentials(apiKey, projectID string) error {
	client := GetClient(apiKey)
	_, _, err := client.Projects.Get(projectID, nil)
	return err
}

func GetClient(apiKey string) *packngo.Client {
	client := packngo.NewClientWithAuth("kubermatic", apiKey, nil)
	client.UserAgent = fmt.Sprintf("kubermatic/api %s", client.UserAgent)
	return client
}

File: modules/web/src/app/core/services/wizard/provider/provider.ts
// Copyright 2020 The Kubermatic Kubernetes Platform contributors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {HttpClient, HttpHeaders} from '@angular/common/http';
import {environment} from '@environments/environment';
import {NodeProvider} from '@shared/model/NodeProviderConstants';

export abstract class Provider {
  private _requiredHeaders = [];

  protected _headers = new HttpHeaders();
  protected _restRoot = environment.restRoot;
  protected _newRestRoot = environment.newRestRoot;
  protected readonly _url = `${this._newRestRoot}/projects/${this._projectID}/providers/${this._provider}/sizes`;

  protected constructor(
    protected _http: HttpClient,
    protected readonly _projectID: string,
    protected readonly _provider: NodeProvider
  ) {}

  protected _setRequiredHeaders(...headers: any): void {
    this._requiredHeaders = headers;
  }

  protected _addRequiredHeader(header: any): void {
    this._requiredHeaders.push(header);
  }

  protected _hasRequiredHeaders(): boolean {
    return (
      this._headers.get(Provider.SharedHeader.Credential) !== null ||
      this._requiredHeaders.filter(header => this._headers.keys().includes(header)).length ===
        this._requiredHeaders.length
    );
  }

  protected _cleanupOptionalHeaders(): void {
    for (const hKey of this._headers.keys()) {
      if (!this._requiredHeaders.includes(hKey)) {
        this._headers = this._headers.delete(hKey);
      }
    }
  }

  protected _credential(credential: string): void {
    if (credential) {
      this._headers = this._headers.set(Provider.SharedHeader.Credential, credential);
    }
  }
}

export namespace Provider {
  export enum SharedHeader {
    Credential = 'Credential',
  }
}
